### ISSUE TITLE
Few build issues on Linux (-lcpprest -lrlclientlib, jdk)

### ISSUE THREAD

**Issue by trufanov-nok** [JohnLangford/vowpal_wabbit#1494]
1. Today I've downloaded a fresh VW ver and tried to build it on my new Kubuntu 18.04 machine. The old one I had before could be built fine on it. But current master branch has some dependency problems:
I cloned project and perform `make && make test`
make test failed with
```
venclude -fPIC  -I ../rapidjson/include -I ../explore -I ../include -I ../rlclientlib -c callback_test.cc -o callback_test.o
In file included from eventhub_client_test.cc:6:0:
http_server/stdafx.h:25:10: fatal error: cpprest/json.h: No such file or directory
 #include "cpprest/json.h"
          ^~~~~~~~~~~~~~~~
compilation terminated.
```
So apparently VW now requires libcpprest-dev for compilation. And this dependency is not documented. 

2. After lib installation with `sudo apt install libcpprest-dev` I tried to `make test` again and it failed again with:

```
g++  concurrent_queue_test.o  str_util_test.o  eventhub_client_test.o  callback_test.o  main.o  ranking_response_test.o  async_batcher_test.o  live_model_test.o  event_test.o  http_server/http_server.o -L ../rlclientlib -lm -pthread -lboost_system -lcrypto -lssl -lcpprest -lrlclient -lboost_unit_test_framework -ldl -Wall -lm -pthread -lboost_system -lcrypto -lssl -lcpprest -lrlclient -lboost_unit_test_framework -ldl -o rlclient-test.out
/usr/bin/x86_64-linux-gnu-ld: cannot find -lrlclient
collect2: error: ld returned 1 exit status
```

So I have no rlclient lib onboard either. I found its sources in `./decision_service/rlclientlib`, but apparently make command haven't been called in this subfolder. I browsed to it, called `make` here and it has been compiled without any problem. After this my `make test` completed without errors and test fails. 

3. Then I tried `make all && make test` and got 

```
In file included from src/main/c++/vowpalWabbit_VW.cc:2:0:
src/main/c++/vowpalWabbit_VW.h:2:10: fatal error: jni.h: No such file or directory
 #include <jni.h>
          ^~~~~~~
compilation terminated.

```
So I had to get jdk with `sudo apt install openjdk-11-jdk`. I think this isn't documented too. Still I had to set JAVA_HOME environment variable visible to `./java/Makefile`. I think the easiest way is to add
`JAVA_HOME := $(shell readlink -f /usr/bin/javac | sed "s:\/bin\/javac::")`
before
`JAVA_INCLUDE = -I $(JAVA_HOME)/include` in it.
This will work only for Linux and doesn't check if JAVA_HOME was already defined outside. But I think it would be a good idea to check if JAVA_HOME is empty and try to detect it at lease for nix machines. And probably echo the jdk used.

**Comment by arielf** [comment]
I have this script to detect java home on Unix-like systems. I call it `java-home`.
It relies only on `readlink` which is part of GNU `coreutils`
```
#!/bin/bash

# Find the absolute path of the java executable
# follow all symlinks (to cover /etc/alternatives/java ...)

jpath=$(readlink -e $(which java))

if [[ "$jpath" != "" ]]; then
    while [[ "$jpath" != "/" ]]; do
        jpath=$(dirname $jpath)
        tryinc="$jpath/include"
        if [[ -d "$tryinc" ]]; then
            echo $jpath
            exit 0
        fi
    done
fi

echo "$0: Couldn't find home dir for java (JAVA_HOME)" >&2
echo "$0: Do you have java installed (developer headers included)?" >&2
exit 1
```

**Comment by JohnLangford** [comment]
I've fixed this in the current master.  Essentially, there is work going on here: https://github.com/JohnLangford/vowpal_wabbit/tree/master/decision_service to create a consumable C++ "reinforcement learning" library powered by VW.  This introduces additional dependencies for things like communicating with a cloud service, but there is no fundamental reason why using VW should require them, so I restructured the build a little bit.
